### PR TITLE
[lliu41] First pull request - app v1

### PR DESCRIPTION
**Pull request by kristol07**
Complete first version of application <bullscows> with tests

### CODE REVIEW

File: BullsAndCows/BullsAndCowsGame.cs [hunk 1]
```diff
@@ -1,20 +1,62 @@
...
+        private readonly string secret;
+        private int leftChance = 6;
         public BullsAndCowsGame(SecretGenerator secretGenerator)
         {
             this.secretGenerator = secretGenerator;
+            this.secret = secretGenerator.GenerateSecret();
         }
 
-        public bool CanContinue => true;
+        public bool CanContinue => leftChance > 0 ? true : false;
```
Comment (FlorianQU) [inline BullsAndCows/BullsAndCowsGame.cs:20]: Could delete true/false expression

File: BullsAndCowsTest/BullsAndCowsGameTest.cs [hunk 2]
```diff
@@ -8,10 +9,183 @@ public class BullsAndCowsGameTest
...
+            //given
+            var mockSecretGenerator = new Mock<SecretGenerator>();
+            mockSecretGenerator.Setup(mock => mock.GenerateSecret()).Returns(secret);
+            var game = new BullsAndCowsGame(mockSecretGenerator.Object);
+
+            //when
+            string answer = game.Guess(guess);
+
+            //then
+            Assert.Equal(bullcows, answer);
```
Comment (Lei-zhu-88) [inline BullsAndCowsTest/BullsAndCowsGameTest.cs:86]: naming: will "expect, actual " are better names?

File: BullsAndCowsTest/BullsAndCowsGameTest.cs [hunk 3]
```diff
@@ -8,10 +9,183 @@ public class BullsAndCowsGameTest
...
+            var secretGenerator = new TestSecretGenerator();
+            var game = new BullsAndCowsGame(secretGenerator);
+
+            string answer = game.Guess(guess);
+
+            Assert.Equal("Wrong Input, input again", answer);
+        }
+
+        [Theory]
+        [InlineData("1 2 23 4")]
```
Comment (Lei-zhu-88) [inline BullsAndCowsTest/BullsAndCowsGameTest.cs:155]: should add cases like "123" "1 s" whose length is not long enough

File: BullsAndCows/BullsAndCowsGame.cs (hunk 1 above)
Comment (kristol07) [reply BullsAndCows/BullsAndCowsGame.cs:20]: agree

File: BullsAndCows/BullsAndCowsGame.cs [hunk 4]
```diff
@@ -1,20 +1,62 @@
...
+
+            CountLeftChance(bulls);
+
+            return $"{bulls}A{cows}B";
+        }
+
+        private void CountLeftChance(int bulls)
+        {
+            this.leftChance =
+                bulls == 4
```
Comment (Lei-zhu-88) [inline BullsAndCows/BullsAndCowsGame.cs:46]: should name number "4" ?

File: BullsAndCowsTest/BullsAndCowsGameTest.cs (hunk 3 above)
Comment (kristol07) [reply BullsAndCowsTest/BullsAndCowsGameTest.cs:155]: `Should_return_wrong_input_tip_when_number_of_digits_is_not_4`  this test method checked this kind of case

File: BullsAndCowsTest/BullsAndCowsGameTest.cs (hunk 2 above)
Comment (kristol07) [reply BullsAndCowsTest/BullsAndCowsGameTest.cs:86]: agree

File: BullsAndCows/BullsAndCowsGame.cs (hunk 4 above)
Comment (kristol07) [reply BullsAndCows/BullsAndCowsGame.cs:46]: agree

Comment (kristol07) [reply BullsAndCows/BullsAndCowsGame.cs:46]: > should name number "4" ?

agree, resolved.

File: BullsAndCowsTest/BullsAndCowsGameTest.cs (hunk 3 above)
Comment (kristol07) [reply BullsAndCowsTest/BullsAndCowsGameTest.cs:155]: > should add cases like "123" "1 s" whose length is not long enough

`Should_return_wrong_input_tip_when_number_of_digits_is_not_4` this test method checked this kind of case

File: BullsAndCowsTest/BullsAndCowsGameTest.cs (hunk 2 above)
Comment (kristol07) [reply BullsAndCowsTest/BullsAndCowsGameTest.cs:86]: > naming: will "expect, actual " are better names?

agree

File: BullsAndCows/BullsAndCowsGame.cs (hunk 1 above)
Comment (kristol07) [reply BullsAndCows/BullsAndCowsGame.cs:20]: > Could delete true/false expression

agree , resolved.

**Pull request closed without merging**